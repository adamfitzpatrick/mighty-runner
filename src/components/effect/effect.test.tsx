import * as React from 'react'
import { cleanup, render, RenderResult, fireEvent } from '@testing-library/react'
import Effect from '.'
import * as Models from '@models'

describe('Effect component', () => {
  let effect: Models.Effect
  let toggleSpy: jest.Mock<Models.Effect>
  let changeSpy: jest.Mock<Models.Effect>
  let blurSpy: jest.Mock<Models.Effect>
  let sut: RenderResult

  beforeEach(() => {
    effect = {
      id: '1',
      name: 'n',
      description: 'd',
      active: true,
      target: [ 'a', 'b' ],
      value: 1
    }
    toggleSpy = jest.fn()
    changeSpy = jest.fn()
    blurSpy = jest.fn()
  })

  describe('with no edit capability', () => {
    beforeEach(() => {
      sut = render(<Effect effect={ effect } onToggle={ toggleSpy } />)
    })

    test('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
    })

    test('should call onToggle method when active checkbox is clicked', () => {
      const input = sut.getByTestId('Active.input.component')
      const event = {
        bubbles: true,
        target: { checked: false }
      }
      fireEvent(input, new MouseEvent('click', event))
      effect.active = false
      expect(toggleSpy).toHaveBeenCalledWith(effect)
    })
  })

  describe('with edit capability', () => {
    beforeEach(() => {
      sut = render(
        <Effect
          effect={ effect }
          onChange={ changeSpy }
          onBlur={ blurSpy }
        />
      )
    })

    test('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
    })

    test('should call the onChange method when a field is modified', () => {
      const input = sut.getByTestId('Name.input.component')
      const event = { target: { value: 'changed' } }
      fireEvent.change(input, event)
      effect.name = 'changed'
      expect(changeSpy).toHaveBeenCalledWith(effect)
    })

    test('should call the onBlur method when an input field is blurred', () => {
      const input = sut.getByTestId('Name.input.component')
      fireEvent.blur(input)
      expect(blurSpy).toHaveBeenCalled()
    })
  })
})
