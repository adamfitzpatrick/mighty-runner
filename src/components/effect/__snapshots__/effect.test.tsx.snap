// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Effect component with edit capability should render correctly 1`] = `"<div data-testid=\\"editable-effect.component\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"n\\"></label><label><span>Description</span><input data-testid=\\"Description.input.component\\" type=\\"text\\" value=\\"d\\"></label><label>Target<input data-testid=\\"0.array.input\\" type=\\"text\\" value=\\"a\\"></label><label><span class=\\"additionalLabel\\">Target</span><input data-testid=\\"1.array.input\\" type=\\"text\\" value=\\"b\\"></label><label><span>Value</span><input data-testid=\\"Value.input.component\\" type=\\"text\\" value=\\"1\\"></label><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div>"`;

exports[`Effect component with no edit capability should render correctly 1`] = `"<div data-testid=\\"non-editable-effect.component\\"><span>Name: n</span><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div>"`;
